File: midonet-api/src/main/java/org/midonet/api/neutron/PortResource.java
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.midonet.api.neutron;

import java.util.List;
import java.util.UUID;

import javax.annotation.security.RolesAllowed;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import javax.ws.rs.core.UriInfo;

import com.google.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.midonet.cluster.auth.AuthRole;
import org.midonet.api.rest_api.AbstractResource;
import org.midonet.api.rest_api.RestApiConfig;
import org.midonet.cluster.rest_api.NotFoundHttpException;
import org.midonet.cluster.rest_api.neutron.NeutronMediaType;
import org.midonet.cluster.rest_api.neutron.NeutronUriBuilder;
import org.midonet.cluster.rest_api.neutron.models.Port;
import org.midonet.cluster.services.rest_api.neutron.plugin.NetworkApi;
import org.midonet.event.neutron.PortEvent;

import static org.midonet.cluster.rest_api.validation.MessageProperty.RESOURCE_NOT_FOUND;
import static org.midonet.cluster.rest_api.validation.MessageProperty.getMessage;

public class PortResource extends AbstractResource {

    private final static Logger log = LoggerFactory.getLogger(
            PortResource.class);
    private final static PortEvent PORT_EVENT =
            new PortEvent();

    private final NetworkApi api;

    @Inject
    public PortResource(RestApiConfig config, UriInfo uriInfo,
                        SecurityContext context, NetworkApi api) {
        super(config, uriInfo, context, null, null);
        this.api = api;
    }

    @POST
    @Consumes(NeutronMediaType.PORT_JSON_V1)
    @Produces(NeutronMediaType.PORT_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public Response create(Port port) {
        log.info("PortResource.create entered {}", port);
        Port p = api.createPort(port);
        PORT_EVENT.create(p.id, p);
        log.info("PortResource.create exiting {}", p);
        return Response.created(
            NeutronUriBuilder.getPort(
                getBaseUri(), p.id)).entity(p).build();
    }

    @POST
    @Consumes(NeutronMediaType.PORTS_JSON_V1)
    @Produces(NeutronMediaType.PORTS_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public Response createBulk(List<Port> ports) {
        log.info("PortResource.createBulk entered");

        List<Port> outPorts = api.createPortBulk(ports);
        for (Port p : outPorts) {
            PORT_EVENT.create(p.id, p);
        }
        return Response.created(NeutronUriBuilder.getPorts(
            getBaseUri())).entity(outPorts).build();
    }

    @DELETE
    @Path("{id}")
    @RolesAllowed(AuthRole.ADMIN)
    public void delete(@PathParam("id") UUID id) {
        log.info("PortResource.delete entered {}", id);
        api.deletePort(id);
        PORT_EVENT.delete(id);
    }

    @GET
    @Path("{id}")
    @Produces(NeutronMediaType.PORT_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public Port get(@PathParam("id") UUID id) {
        log.info("PortResource.get entered {}", id);

        Port p = api.getPort(id);
        if (p == null) {
            throw new NotFoundHttpException(getMessage(RESOURCE_NOT_FOUND));
        }

        log.info("PortResource.get exiting {}", p);
        return p;
    }

    @GET
    @Produces(NeutronMediaType.PORTS_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public List<Port> list() {
        log.info("PortResource.list entered");
        return api.getPorts();
    }

    @PUT
    @Path("{id}")
    @Consumes(NeutronMediaType.PORT_JSON_V1)
    @Produces(NeutronMediaType.PORT_JSON_V1)
    @RolesAllowed(AuthRole.ADMIN)
    public Response update(@PathParam("id") UUID id, Port port) {
        log.info("PortResource.update entered {}", port);

        Port p = api.updatePort(id, port);
        PORT_EVENT.update(id, p);
        log.info("PortResource.update exiting {}", p);
        return Response.ok(
            NeutronUriBuilder.getPort(
                getBaseUri(), p.id)).entity(p).build();

    }
}

File: cluster/midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/PortResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.{List => JList, UUID}
import javax.ws.rs._
import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response
import javax.ws.rs.core.Response.Status
import javax.ws.rs.core.Response.Status._

import scala.collection.JavaConverters._
import scala.concurrent.Await
import scala.reflect.ClassTag
import scala.util.control.NonFatal

import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.rest_api.ResponseUtils._
import org.midonet.cluster.rest_api.annotation._
import org.midonet.cluster.rest_api.models.Route.NextHop
import org.midonet.cluster.rest_api.models._
import org.midonet.cluster.rest_api.validation.MessageProperty._
import org.midonet.cluster.rest_api.{InternalServerErrorHttpException, NotFoundHttpException}
import org.midonet.cluster.services.MidonetBackend.HostsKey
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.{Create, ResourceContext, Update}
import org.midonet.cluster.util.SequenceType

class AbstractPortResource[P >: Null <: Port] (resContext: ResourceContext)
                                              (implicit tag: ClassTag[P])
    extends MidonetResource[P](resContext)(tag) {

    protected override def getFilter = (port: P) => setActive(port)

    protected override def listFilter = (port: P) => { setActive(port); true }

    private def isActive(id: String): Boolean = {
        getResourceState(classOf[Port], id, HostsKey).getOrThrow.nonEmpty
    }

    private def setActive(port: P): P = {
        port.active = isActive(port.id.toString)
        port
    }

    protected[resources] def ensureTunnelKey(port: P): Unit = try {
        port.tunnelKey = Await.result (
            resContext.seqDispenser.next(SequenceType.OverlayTunnelKey),
            MidonetResource.Timeout
        ).toInt
    } catch {
        case NonFatal(t) =>
            log.error("Failed to generate tunnel key", t)
            throw new InternalServerErrorHttpException(
                s"Unable to generate new tunnel key: ${t.getMessage}")
    }

}

@RequestScoped
@AllowGet(Array(APPLICATION_PORT_JSON,
                APPLICATION_PORT_V2_JSON,
                APPLICATION_JSON))
@AllowList(Array(APPLICATION_PORT_COLLECTION_JSON,
                 APPLICATION_PORT_V2_COLLECTION_JSON,
                 APPLICATION_JSON))
@AllowUpdate(Array(APPLICATION_PORT_JSON,
                   APPLICATION_PORT_V2_JSON,
                   APPLICATION_JSON))
@AllowDelete
class PortResource @Inject()(resContext: ResourceContext)
    extends AbstractPortResource[Port](resContext) {

    @POST
    @Path("{id}/link")
    @Consumes(Array(APPLICATION_PORT_LINK_JSON,
                    APPLICATION_JSON))
    def link(@PathParam("id") id: UUID, link: Link): Response = {
        getResource(classOf[Port], id) flatMap { port =>
            getResource(classOf[Port], link.peerId) map { peerPort =>
                if ((port.peerId eq null) && (peerPort.peerId eq null)) {
                    port.peerId = link.peerId
                    updateResource(port, Response.created(link.getUri).build())
                } else {
                    buildErrorResponse(BAD_REQUEST.getStatusCode,
                                       getMessage(PORTS_LINKABLE))
                }
            }
        } getOrThrow
    }

    @DELETE
    @Path("{id}/link")
    def unlink(@PathParam("id") id: UUID): Response = {
        getResource(classOf[Port], id) map { port =>
            port.peerId = null
            updateResource(port, MidonetResource.OkNoContentResponse)
        } getOrThrow
    }

    @Path("{id}/port_groups")
    def portGroups(@PathParam("id") id: UUID): PortPortGroupResource = {
        new PortPortGroupResource(id, resContext)
    }

    // The final is not totally required, but it just helps confirming that
    // nobody is overriding the tunnelKey check below
    protected final override def updateFilter = (to: Port, from: Port) => {
        to.tunnelKey = from.tunnelKey // disallow updating
        to.update(from)
    }

}

@RequestScoped
@AllowList(Array(APPLICATION_PORT_COLLECTION_JSON,
                 APPLICATION_PORT_V2_COLLECTION_JSON,
                 APPLICATION_JSON))
@AllowCreate(Array(APPLICATION_PORT_JSON,
                   APPLICATION_PORT_V2_JSON,
                   APPLICATION_JSON))
class BridgePortResource @Inject()(bridgeId: UUID,
                                   resContext: ResourceContext)
    extends AbstractPortResource[BridgePort](resContext) {

    protected override def listFilter = (port: Port) => {
        port.getDeviceId == bridgeId
    }

    protected override def createFilter = (port: BridgePort) => {
        ensureTunnelKey(port)
        port.create(bridgeId)
    }
}

@RequestScoped
@AllowList(Array(APPLICATION_PORT_COLLECTION_JSON,
                 APPLICATION_PORT_V2_COLLECTION_JSON,
                 APPLICATION_JSON))
class RouterPortResource @Inject()(routerId: UUID, resContext: ResourceContext)
    extends AbstractPortResource[RouterPort](resContext) {

    protected override def listFilter = (port: Port) => {
        port.getDeviceId == routerId
    }

    @POST
    @Consumes(Array(APPLICATION_PORT_JSON,
                    APPLICATION_PORT_V2_JSON,
                    APPLICATION_JSON))
    override def create(port: RouterPort,
                        @HeaderParam("Content-Type") contentType: String)
    : Response = {

        throwIfViolationsOn(port)

        ensureTunnelKey(port)
        port.create(routerId)
        port.setBaseUri(resContext.uriInfo.getBaseUri)
        val route = new Route("0.0.0.0", 0, port.portAddress, 32, NextHop.Local,
                              port.id, "255.255.255.255", 0, routerId, false)
        route.setBaseUri(resContext.uriInfo.getBaseUri)
        multiResource(Seq(Create(port), Create(route)),
                      Response.created(port.getUri).build())
    }

}

@RequestScoped
class BridgePeerPortResource @Inject()(routerId: UUID,
                                       resContext: ResourceContext)
    extends AbstractPortResource[Port](resContext) {

    @GET
    @Produces(Array(APPLICATION_PORT_COLLECTION_JSON,
                    APPLICATION_PORT_V2_COLLECTION_JSON,
                    APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String): JList[Port] = {
        (getResource(classOf[Bridge], routerId) flatMap { router =>
            listResources(classOf[Port], router.portIds.asScala)
        } flatMap { ports =>
            val peerPortIds = ports.filter(_.peerId ne null).map(_.peerId)
            listResources(classOf[Port], peerPortIds)
         } getOrThrow).asJava
    }

}

@RequestScoped
class RouterPeerPortResource @Inject()(routerId: UUID,
                                       resContext: ResourceContext)
    extends AbstractPortResource[Port](resContext) {

    @GET
    @Produces(Array(APPLICATION_PORT_COLLECTION_JSON,
                    APPLICATION_PORT_V2_COLLECTION_JSON,
                    APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String): JList[Port] = {
        (getResource(classOf[Router], routerId) flatMap { router =>
            listResources(classOf[Port], router.portIds.asScala)
        } flatMap { ports =>
            val peerPortIds = ports.filter(_.peerId ne null).map(_.peerId)
            listResources(classOf[Port], peerPortIds)
         } getOrThrow).asJava
    }

}

@RequestScoped
class PortGroupPortResource @Inject()(portGroupId: UUID,
                                      resContext: ResourceContext)
    extends MidonetResource[PortGroupPort](resContext) {

    @GET
    @Produces(Array(APPLICATION_PORTGROUP_PORT_JSON,
                    APPLICATION_JSON))
    @Path("{id}")
    override def get(@PathParam("id") id: String,
                     @HeaderParam("Accept") accept: String): PortGroupPort = {
        val portId = UUID.fromString(id)
        val portGroupPort = getResource(classOf[PortGroupPort], portId)
            .getOrThrow
        if (!portGroupPort.portGroupIds.contains(portGroupId)) {
            throw new NotFoundHttpException("Resource not found")
        }
        portGroupPort.portGroupId = portGroupId
        portGroupPort
    }

    @GET
    @Produces(Array(APPLICATION_PORTGROUP_PORT_COLLECTION_JSON,
                    APPLICATION_JSON))
    override def list(@HeaderParam("Accept") accept: String)
    : JList[PortGroupPort] = {
        val portGroupPorts = getResource(classOf[PortGroup], portGroupId)
            .flatMap(pg => listResources(classOf[PortGroupPort],
                                         pg.portIds.asScala))
            .getOrThrow
        portGroupPorts.foreach(portGroupPort => {
            if (!portGroupPort.portGroupIds.contains(portGroupId)) {
                throw new NotFoundHttpException("Resource not found")
            }
            portGroupPort.portGroupId = portGroupId
        })
        portGroupPorts.asJava
    }

    @POST
    @Consumes(Array(APPLICATION_PORTGROUP_PORT_JSON,
                    APPLICATION_JSON))
    override def create(portGroupPort: PortGroupPort,
                        @HeaderParam("Content-Type") contentType: String)
    : Response = {
        portGroupPort.portGroupId = portGroupId

        throwIfViolationsOn(portGroupPort)

        getResource(classOf[Port], portGroupPort.portId).flatMap(port => {
            getResource(classOf[PortGroup], portGroupPort.portGroupId).map(pg => {
                if (port.portGroupIds.contains(portGroupPort.portGroupId)) {
                    Response.status(Status.CONFLICT).build()
                } else  if (pg.portIds.contains(portGroupPort.portId)) {
                    Response.status(Status.CONFLICT).build()
                } else {
                    portGroupPort.setBaseUri(resContext.uriInfo.getBaseUri)
                    port.portGroupIds.add(portGroupPort.portGroupId)
                    pg.portIds.add(portGroupPort.portId)
                    multiResource(Seq(Update(port), Update(pg)),
                                  Response.created(portGroupPort.getUri)
                                      .build())
                }
            })
        }).getOrThrow
    }

    @DELETE
    @Path("{id}")
    override def delete(@PathParam("id") id: String): Response = {
        val portId = UUID.fromString(id)
        // If the port still exists, delete the port group ID from the port,
        // otherwise delete only the port ID from the port group.
        getResource(classOf[Port], portId).flatMap(port => {
            getResource(classOf[PortGroup], portGroupId).map(pg => {
                if (!port.portGroupIds.contains(portGroupId)) {
                    Response.status(Status.NOT_FOUND).build()
                } else if (!pg.portIds.contains(portId)) {
                    MidonetResource.OkNoContentResponse
                } else {
                    port.portGroupIds.remove(portGroupId)
                    pg.portIds.remove(portId)
                    multiResource(Seq(Update(port), Update(pg)),
                                  MidonetResource.OkNoContentResponse)
                }
            })
        }).fallbackTo(
            getResource(classOf[PortGroup], portGroupId).map(pg => {
                if (!pg.portIds.contains(portId)) {
                    MidonetResource.OkNoContentResponse
                } else {
                    pg.portIds.remove(portId)
                    updateResource(pg, MidonetResource.OkNoContentResponse)
                }
            }))
        .getOrThrow
    }
}